arch/risc-v/src/common/riscv_mtimer.c

```diff
@@ -0,0 +1,119 @@
+/****************************************************************************
+ * arch/risc-v/src/common/riscv_mtimer.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+
+#include <arch/barriers.h>
+
+#include <stdint.h>
+
+#include "riscv_internal.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define is_lt_u64(a, b)     ((int64_t)((int64_t)(a) - (int64_t)(b)) < 0)
+#define is_gtoreq_u64(a, b) !is_lt_u64(a, b)
+
+#ifdef CONFIG_ARCH_RV32
+#  warning "Missing logic... Access to 64-bit registers needs fixing"
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* Memory mapped register base addresses */
+
+static volatile uintptr_t g_mtimecmp_base;
+static volatile uintptr_t g_mtime;
+
+/* Amount of M-mode ticks for one OS tick */
+
+static volatile uint64_t  g_tick;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void riscv_init_mtimer(uintptr_t mcmpbase, uintptr_t mtime, uint64_t tick)
```

Comment:
I simply moved what was in mpfs_timerisr here and I think quite many other RISC-V targets implement the same timer, one other example here: https://github.com/apache/incubator-nuttx/blob/master/arch/risc-v/src/bl602/bl602_timerisr.c

I did not know of this oneshot/alarm functionality, but this driver could be quite easily converted into an alarm timer, just move the tick argument out and give the next alarm time in. This driver would then handle the upper half of the ISR and dispatch / ask for the lower half to execute.

I did not plan on doing any of this, I just wanted to move what was in mpfs_timerisr to a more visible location, because there was no MPFS specific code there (90% copy&paste, 10% MPFS).

Reply:
Ok, we have write a generic mtimer driver before, please wait me clean up and upstream by today. You can try it with mpfs.

Reply:
Ok, let's see how your timer fits.

A last note: I used this implementation as the timer driver for the nuttx SBI (with modifications of course) but maybe your timer fits for that too!

Reply:
@pussuw here is https://github.com/apache/incubator-nuttx/pull/6020, please review and try it. Sorry, it's a little bit delay.

Reply:
Thanks, I'll do the review on monday

Reply:
> Thanks, I'll do the review and test it on monday

thanks.


> Ok, let's see how your timer fits.
> 
> A last note: I used this implementation as the timer driver for the nuttx SBI (with modifications of course) but maybe your timer fits for that too!

It's better to simulate S-mode timer with wdog interface since it's more simpler and could save one hardware timer.

Reply:
For NuttX using some different timer is fine.

But for SBI this is not an option, the SBI actually takes ownership of mtimer in this case, because it is one of its specified responsibilities. A driver is needed for this, but the driver can be quite simple and I can move this simple implementation into the NuttX native SBI, so it does not get mixed with your driver that is more flexible.